### DISCUSSION TITLE
Potential bug with reset?

### DISCUSSION THREAD

**Discussion by terebentina** [react-hook-form/react-hook-form#8138] [General] [answered]
Hi,

I have a problem with `reset` in the latest version (7.29.0). Basically my `defaultValues` contain a field (say A) and later on the data from db doesn't have the said field A. I expect that when resetting to the new data from db in `useEffect`, A shouldn't be present anymore but it still is!

I created a very simple CSB to show this: https://codesandbox.io/s/priceless-moser-4sehuk?file=/src/App.js
Basically, I do
```
useForm({ defaultValues: {a: 'a', b: 'b'} });
```
and then onClick on some button do `reset({ b: 'b2' }, { keepDefaultValues: false, keepValues: false })`. I would expect the submitted data to not include `a` anymore but it does...
I even tried `reset({ a: undefined, b: 'b2' }, { keepDefaultValues: false, keepValues: false })` and still get `a` when submitting.

Did I stumble upon a bug or I am missing something obvious?

**Comment by ritikbanger** [comment, marked as answer]
You are not unregistering the fields. The fields are still registered.

**Reply by bluebill1049** [reply]
That's correct. register run again on each render. thanks @ritikbanger 🙏

**Reply by terebentina** [reply]
Reading some more on `reset`, I think the confusion comes from the fact that, currently, `reset` does a partial update and I was expecting it to do a full (hard?) reset of the entire form. Intuitively I thought `keepDefaultValues: false, keepValues: false` would help me do that.
Basically what I need is something like this:
1. `unregister` everything (both `a` and `b`)
2. `reset` to the new values (`{ b: 'new value' }`)
3. let `register` bring the refs back on re-render so the form values are now `{ a: undefined, b: 'new value' }`

To me this looks like too common a scenario to be run in userland...am I the only one doing this? 😁
Something like `reset({ ... }, { removeMissingValues: true })` or `reset({ ... }, { fullReset: true })`

I guess my alternative, for now, is to manually check which fields are missing from the backend payload and set them to the empty string.
Do you see a better approach?

**Reply by bluebill1049** [reply]
In the next major version, useForm and reset will be required to provide complete `defaultValues` instead of partial. It's difficult to deep merge and even guess what's the defaultValue for the missing values.

**Reply by terebentina** [reply]
Bill, that makes a lot of sense actually and I bet it'll simplify rhf a lot!
